### PR TITLE
Templating to be able to use smaller type

### PR DESCRIPTION
**Pull request by stephanecollot**
This MR allows to use float32 instead of the default float64.

In this MR:
- Using C++ templating with float (float32) and double (float64)
- Using Cython templating with `fused`
- Automatically use the right implementation depending on matrix input type
- Update readme
- Adding unit test to test both types
- Adding unit test to cover ArrayWrapper_*

### CODE REVIEW

**Comment by stephanecollot** [comment]
Hi @ParticularMiner 

I just rebased this MR on top of yours.
Everything is working, except the ArrayWrapper part.
The Unit test:
_tests/sparse_dot_topn/test_awesome_cossim_topn.py::test_awesome_cossim_top_small_matrix[**float32**]_
fails with:
```
>       return array(a, dtype, copy=False, order=order)
E       RuntimeError: Item size 4 for PEP 3118 buffer format string d does not match the dtype d item size 8.

../../../miniconda3/envs/py3/lib/python3.7/site-packages/numpy/core/_asarray.py:102: RuntimeError
```

I was wondering, could you have a look? I can't see why it is failing some how.

If you want to see only my modification, since your branch is not yet merged, have a look here: https://github.com/stephanecollot/sparse_dot_topn/pull/1/files

**Comment by ParticularMiner** [comment]
Sure @stephanecollot 

I'll take a look at it.

**Comment by ParticularMiner** [comment]
Notify: @stephanecollot 

Do you by any chance get the following compilation warning?

>    ./sparse_dot_topn\sparse_dot_topn_threaded.cpp(3174): warning C4244: '=': conversion from 'long' to 'char', possible loss of data
    ./sparse_dot_topn\sparse_dot_topn_threaded.cpp(5742): warning C4244: '=': conversion from 'long' to 'char', possible loss of data
    ./sparse_dot_topn\sparse_dot_topn_threaded.cpp(8275): warning C4244: '=': conversion from 'long' to 'char', possible loss of data

File: sparse_dot_topn/array_wrappers.pyx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    def __getbuffer__(self, Py_buffer *buffer, int flags):
+        # relevant documentation http://cython.readthedocs.io/en/latest/src/userguide/buffer.html#a-matrix-class
+        cdef Py_ssize_t itemsize = sizeof(self.vec[0])
+
+        self.shape[1] = self.vec.size()
+        self.shape[0] = 1
+        self.strides[1] = <Py_ssize_t>(  <char *>&(self.vec[1]) - <char *>&(self.vec[0]))
+        self.strides[0] = self.vec.size() * self.strides[1]
+        buffer.buf = <char *>&(self.vec[0])
+        buffer.format = 'd'
```
Comment (ParticularMiner) [inline sparse_dot_topn/array_wrappers.pyx]: Hi @stephanecollot 

All the unit-tests passed (and the compilation warnings vanished) after changing this line from 
```python
buffer.format = 'd'
```
to
```python
buffer.format = 'f'
``` 

Do let me know if you encounter further problems after you effect this change.

**Comment by stephanecollot** [comment]
@ParticularMiner I don't get any warning like that in the console when I run:

```
python setup.py clean --all
python setup.py develop
```

**Comment by ParticularMiner** [comment]
@stephanecollot 

Then it's probably platform-dependent.  I'm using Visual C++ compiler on Windows 10.


> @ParticularMiner I don't get any warning like that in the console when I run:
> 
> 
> 
> ```
> 
> python setup.py clean --all
> 
> python setup.py develop
> 
> ```

File: sparse_dot_topn/array_wrappers.pyx (hunk 1 above)
Comment (stephanecollot) [reply sparse_dot_topn/array_wrappers.pyx]: Nice thanks good catch! I pushed it.

**Comment by stephanecollot** [comment]
Thanks for the help.

I just pushed more unit test.

**Comment by stephanecollot** [comment]
I just rebased on master, it is ready* to release @ymwdalex

**Comment by ParticularMiner** [comment]
@stephanecollot @ymwdalex @RUrlus

It was my pleasure to contribute to this project.  I wish you all the best!


> Thanks for the help.
> 
> I just pushed more unit test.

**Review by mbaak — APPROVED** [review]

**Comment by mbaak** [comment]
@stephanecollot I went through the changes, looks okay to me.

**Comment by ParticularMiner** [comment]
Kindly drop me a line after the new version is updated on pypi.org &mdash; I'd like to incorporate the changes in our application [`string_grouper`](https://github.com/Bergvca/string_grouper) as soon as possible.

😄

**Comment by ymwdalex** [comment]
@stephanecollot I push a new commit on master branch: update the changelog.md, resolved the release problem.

Could you please rebase this branch on top of master, and change the CHANGELOG.md and version number in setup.py? Thanks!

**Review by ymwdalex — APPROVED** [review]
LGTM. Thanks!

**Comment by ymwdalex** [comment]
v0.3.1 is released.  Thanks for your contribution, @stephanecollot  and @ParticularMiner 
https://pypi.org/project/sparse-dot-topn/0.3.1/

**Comment by ParticularMiner** [comment]
Hi @stephanecollot , @ymwdalex, @RUrlus

FYI, here are a few timings obtained on my modest laptop (Intel Core i7, 2.20 GHz clock speed, 16GB RAM).

Clearly, it was a great idea to try out the `float32` datatype!  In my opinion, the precision offered by `float64` is often unnecessarily high for NLP applications.  Fortunately `TfidfVectorizer` from `sklearn.feature_extraction.text` allows the user to specify the datatype of the matrix components.

### Table of average time-durations of `awesome_cossim_topn` (for the old and new versions) over 16 random samples of matrix-pairs ***A*** and ***B***.

|#threads| **(v0.2.9)<br>`float64`**|**(v0.3.0)<br>`float64`**|**(v0.3.1)<br>`float32`**|**%-age speed-increase<br>(v0.3.0 over v0.2.9)**|**%-age speed-increase<br>(`float32` \[v0.3.1\] over `float64` \[v0.3.0\])**|
|---:|---:|---:|---:|---:|---:|
|0|48.357178|45.756128|31.243680|6%|46%|
|1|84.805013|49.046593|32.823485|73%|49%|
|2|58.365927|26.116717|19.529827|123%|34%|
|3|48.586614|21.365131|14.963700|127%|43%|
|4|44.523800|18.311739|12.908124|143%|42%|
|5|42.053307|15.756574|11.992907|166%|31%|
|6|40.212938|14.098908|10.606070|185%|33%|
|7|39.031401|13.224633|9.834058|195%|34%|

#matrix_pairs = 16
ntop = 4&nbsp;000
threshold = 0.01
nr_vocab = 26<sup>3</sup> = 17&nbsp;576 (the number of columns of ***A*** = number of rows of ***B***)
density = 30/n_vocab = 0.0017
n_samples = 1&nbsp;000&nbsp;000 (the number of rows of ***A***)
n_duplicates = 4&nbsp;000  (the number of columns of ***B***)
nnz(***A***) = 30&nbsp;000&nbsp;000  (the number of nonzero components of ***A***)
nnz(***B***) = 120&nbsp;000  (the number of nonzero components of ***B***)

nnz(***A***\****B***) = 188&nbsp;446&nbsp;798 &pm; 44&nbsp;679   (mean and standard deviation of the total number of nonzero components of ***A***\****B*** above the threshold over 16 random matrix-pairs ***A*** and ***B***)

ntop(***A***\****B***) = 391 &pm; 15  (mean and standard deviation of the true maximum number of nonzero components per row of ***A***\****B*** above the threshold over 16 random matrix-pairs ***A*** and ***B***)

**Comment by stephanecollot** [comment]
Nice, I think we are going to make a blog article about it.
@ParticularMiner let me know if you want to be mentioned and how, contact me here: stephane  dot  collot   at  gmail  dot  com

**Comment by stephanecollot** [comment]
@ParticularMiner here is the draft article: https://docs.google.com/document/d/1QLSoSAf6gncJ27J9UQ5vN4Zx3Na6H92gihQ0DCb4l9s/edit#
Feel free to edit

**Comment by ParticularMiner** [comment]
Very kind of you @stephanecollot!  I will soon contact you by email.

**Pull request merged**